utils/disk_space_monitor.hh

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <chrono>
+#include <filesystem>
+
+#include <boost/signals2/connection.hpp>
+#include <boost/signals2/signal_type.hpp>
+#include <boost/signals2/dummy_mutex.hpp>
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include "seastarx.hh"
+#include "utils/assert.hh"
+#include "utils/log.hh"
+
+using namespace std::chrono_literals;
+namespace bs2 = boost::signals2;
+
+namespace utils {
+
+extern seastar::logger dlmlog;
+
+// Instantiated only on shard 0
+template <typename Clock = lowres_clock>
+class disk_space_monitor {
+public:
+    using signal_type = bs2::signal_type<void (), bs2::keywords::mutex_type<bs2::dummy_mutex>>::type;
+    using signal_callback_type = std::function<future<>(const disk_space_monitor&)>;
+    using signal_connection_type = bs2::scoped_connection;
+
+    struct config {
+        scheduling_group sched_group;
+        Clock::duration normal_polling_interval = 10s;
+        Clock::duration high_polling_interval = 1s;
+        // Use high_polling_interval above this threshold
+        float polling_interval_threshold = 0.9;
+    };
+
+private:
+    abort_source _as;
+    optimized_optional<abort_source::subscription> _as_sub;
+    future<> _poller_fut = make_ready_future();
+    signal_type _signal_source;
+    uint64_t _total_space = 0;
+    uint64_t _free_space = 0;
+
+protected:
+    config _cfg;
+    std::filesystem::path _data_dir;
+
+    virtual future<struct statvfs> get_filesystem_stats() const {
+        return engine().statvfs(_data_dir.native());
+    }
+public:
+    explicit disk_space_monitor(abort_source& as, config cfg)
+        : _as_sub(as.subscribe([this] () noexcept { _as.request_abort(); }))
+        , _cfg(std::move(cfg))
+    {}
+    virtual ~disk_space_monitor() {
+        SCYLLA_ASSERT(_poller_fut.available());
+    }
+
+    void start(std::filesystem::path data_dir) {
+        _data_dir = std::move(data_dir);
+        _poller_fut = poll();
```

Comment:
nit: we could have passed data dir in the configuration.

Reply:
yes, or pass it to the constructor